pkg/connection/connection.go

```diff
@@ -0,0 +1,83 @@
+package connection
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"io"
+	"net/http"
+	"time"
+)
+
+// Connection is to be implemented by any struct that attempts to perform
+// requests against a remote resource that implements the /connect API.
+type Connection interface {
+	// Returns an HTTP request object that can be used by a subsequent `Do`
+	// call.
+	GetRequest(string, string, any) (*http.Request, error)
+
+	// Performs an HTTP request to the remote API. Returns the response body or
+	// an error object.
+	Do(*http.Request) ([]byte, error)
```

Comment:
Mhm, do we need the context outside of the actual requesting function is the question. In most cases we just want to check if everything was alright and move on.
If I would return `http.Request` instead of a handcrafted struct or?

Reply:
In my mind we only need to return the body. For the rest of the fields on this hypothetical `Response` object:

1. `StatusCode`: sounds like a good idea, but in the end callers of `Do` want to perform a request and get some to be parsed as a JSON value in return. And, if that is not returned, then callers expect an `error`. This means that 401, 404, etc. are to be handled internally in `Do`, and then the caller can simply print the error value and be done with it. From the perspective of the caller, things are easy with this scheme.
2. `Headers`: the only header I see people caring about is the one containing the new system token, but that is already handled through the code on `Do`. That is, notice that `TODO` in there which says to handle the system token. In the end, this would be a matter of extracting the header inside of `Do`, and then calling `Credentials.Update` with the new value. Hence, callers of `Do` shouldn't need to care about headers because the relevant ones are already handled automatically (or that is the intention, I think :P)
3. `Request`: this is already available because the caller actually used it to call `Do`.

Reply:
understood. I though of it from a common http request service.